scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -42,37 +55,102 @@ class KRouter[A](val config: Config[A], val network: KNetwork[A])(
       )
   }
 
-  // Handle findNodes requests...
-  val nodes: ConnectableObservable[(FindNodes[A], Nodes[A] => Task[Unit])] = network.findNodes
-  nodes.foreach {
-    case (findNodesRequest, responseHandler) =>
-      val FindNodes(uuid, nodeRecord, targetNodeId) = findNodesRequest
-
+  network.kRequests.foreach {
+    case (FindNodes(uuid, nodeRecord, targetNodeId), responseHandler) =>
       debug(
         s"Received request FindNodes(${nodeRecord.id.toHex}, $nodeRecord, ${targetNodeId.toHex})"
       )
       add(nodeRecord)
-      val result = embellish(kBuckets.closestNodes(targetNodeId, config.k))
-      responseHandler(Nodes(uuid, config.nodeRecord, result)).runToFuture
-        .onComplete {
-          case Failure(t) =>
-            log.info(
-              s"Nodes response $uuid to ${nodeRecord.routingAddress} failed with exception: $t"
-            )
-          case _ =>
-        }
+
+      val result = Nodes(uuid, config.nodeRecord, embellish(kBuckets.closestNodes(targetNodeId, config.k)))
+
+      sendResponse(result, responseHandler, nodeRecord.routingAddress)
+    case (Ping(uuid, nodeRecord), responseHandler) =>
+      debug(
+        s"Received request Ping(${nodeRecord.id.toHex}, $nodeRecord)"
+      )
+      add(nodeRecord)
+      sendResponse(Pong(uuid, config.nodeRecord), responseHandler, nodeRecord.routingAddress)
   }
-  nodes.connect()
 
   def get(key: BitVector): Future[NodeRecord[A]] = {
     debug(s"get(${key.toHex})")
     getLocally(key).recoverWith { case _ => getRemotely(key) }.recoverWith { case t => giveUp(key, t) }
   }
 
-  def add(nodeRecord: NodeRecord[A]): Unit = {
-    kBuckets.add(nodeRecord.id)
-    nodeRecords.put(nodeRecord.id, nodeRecord)
-    debug(s"Added record (${nodeRecord.id.toHex}, $nodeRecord) to the routing table.")
+  def add(nodeRecord: NodeRecord[A]): Ack = {
```

Comment:
That is blocking, isn't it? You can keep it that way, but please add  comment - that this method won't return until ping-and-add procedure is finished.

Reply:
I understand that this is not blocking (as Ack extends Future[Ack]). However, wasn't there a convention that we should not call `onNext` until the Future returns? Or this only applies to certain overloaded methods that explicitly return `Future[Ack]`?

Reply:
Ack can be used where Future[Ack] is needed, but not the other way, so we can expect it to return synchronously. But you are correct that it is not blocking, there is a buffer, infinite by default. Disregard my first comment.

Reply:
> Ack can be used where Future[Ack] is needed, but not the other way (we cannot return Future[Ack] where Ack is required)

Silly me, you are right.